File: worker/python/dlq.py
"""
isolane/worker/python/dlq.py

Dead-letter queue routing for orphaned work items.
"""

import os
import time
from typing import Any

import redis

MAX_CLAIM_ATTEMPTS = int(os.environ.get("MAX_CLAIM_ATTEMPTS", 3))


def get_claim_count(fields: dict[str, str]) -> int:
    """Extract the _claim_count field from a work item."""
    try:
        return int(fields.get("_claim_count", 0))
    except (ValueError, TypeError):
        return 0


def should_route_to_dlq(fields: dict[str, str]) -> bool:
    """Return True if this item has exceeded the max claim attempts."""
    return get_claim_count(fields) >= MAX_CLAIM_ATTEMPTS


def route_to_dlq(
    r:            redis.Redis,
    namespace:    str,
    stream_id:    str,
    fields:       dict[str, str],
    reason:       str = "max_claims_exceeded",
) -> None:
    """
    Write a work item to the namespace DLQ stream.
    """
    dlq_key = f"{namespace}.dlq"

    dlq_fields = {
        "original_stream_id": stream_id,
        "namespace":          namespace,
        "pipeline_id":        fields.get("pipeline_id", "unknown"),
        "reason":             reason,
        "claim_count":        str(get_claim_count(fields)),
        "routed_at":          str(int(time.time() * 1000)),
    }

    r.xadd(dlq_key, dlq_fields, maxlen=10000, approximate=True)

    print(
        f"[dlq] {namespace} | message {stream_id} routed to DLQ "
        f"after {get_claim_count(fields)} claims | reason: {reason}"
    )

    # Increment Prometheus counter
    try:
        from worker.python.metrics import dlq_total
        dlq_total.labels(namespace=namespace).inc()
    except Exception:
        pass


def increment_claim_count(fields: dict[str, str]) -> dict[str, str]:
    """
    Return a copy of fields with _claim_count incremented.
    """
    updated = dict(fields)
    updated["_claim_count"] = str(get_claim_count(fields) + 1)
    return updated

File: worker/go/dlq.go
// isolane/worker/go/dlq.go
//
// Dead-letter queue routing for the Go worker.
// Matches the Python dlq.py behaviour exactly.

package worker

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"
)

const DefaultMaxClaimAttempts = 3

// GetClaimCount extracts _claim_count from work item fields.
func GetClaimCount(fields map[string]string) int {
	v, ok := fields["_claim_count"]
	if !ok {
		return 0
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return 0
	}
	return n
}

// ShouldRouteToDLQ returns true if the item has exceeded max claim attempts.
func ShouldRouteToDLQ(fields map[string]string, maxAttempts int) bool {
	return GetClaimCount(fields) >= maxAttempts
}

// IncrementClaimCount returns a copy of fields with _claim_count incremented.
func IncrementClaimCount(fields map[string]string) map[string]string {
	updated := make(map[string]string, len(fields))
	for k, v := range fields {
		updated[k] = v
	}
	updated["_claim_count"] = strconv.Itoa(GetClaimCount(fields) + 1)
	return updated
}

// RouteToDLQ writes the work item to the namespace DLQ stream.
func RouteToDLQ(
	ctx       context.Context,
	rdb       redis.UniversalClient,
	namespace string,
	streamID  string,
	fields    map[string]string,
	reason    string,
) error {
	dlqKey := fmt.Sprintf("%s.dlq", namespace)

	dlqFields := map[string]interface{}{
		"original_stream_id": streamID,
		"namespace":          namespace,
		"pipeline_id":        fields["pipeline_id"],
		"reason":             reason,
		"claim_count":        strconv.Itoa(GetClaimCount(fields)),
		"routed_at":          strconv.FormatInt(time.Now().UnixMilli(), 10),
	}

	err := rdb.XAdd(ctx, &redis.XAddArgs{
		Stream: dlqKey,
		MaxLen: 10000,
		Approx: true,
		Values: dlqFields,
	}).Err()

	if err != nil {
		return fmt.Errorf("DLQ write failed: %w", err)
	}

	fmt.Printf(
		"[dlq] %s | message %s routed to DLQ after %d claims | reason: %s\n",
		namespace, streamID, GetClaimCount(fields), reason,
	)

	DLQTotal.WithLabelValues(namespace).Inc()
	return nil
}